app/Domain/Experience/ExperienceAggregateRoot.php

```diff
@@ -0,0 +1,166 @@
+use App\Domain\Experience\Commands\RegisterPullRequest;
+use App\Domain\Experience\Commands\RegisterVideoCompletion;
+use App\Domain\Experience\Commands\UnlockAchievement;
+use App\Domain\Experience\Events\AchievementUnlocked;
+use App\Domain\Experience\Events\ExperienceEarned;
+use App\Domain\Experience\Events\PullRequestMerged;
+use App\Domain\Experience\Events\SeriesCompleted;
+use App\Domain\Experience\Events\VideoCompleted;
+use App\Models\Series;
+use App\Models\User;
+use App\Models\Video;
+use Spatie\EventSourcing\AggregateRoots\AggregateRoot;
+
+class ExperienceAggregateRoot extends AggregateRoot
+{
+    private int $experienceCount = 0;
+
+    private int $pullRequestCount = 0;
+
+    private PullRequestAchievementUnlocker $pullRequestAchievementUnlocker;
+
+    private ExperienceAchievementUnlocker $experienceAchievementUnlocker;
+
+    private SeriesCompletionAchievementUnlocker $seriesCompletionAchievementUnlocker;
+
+    public function __construct()
+    {
+        $this->pullRequestAchievementUnlocker = new PullRequestAchievementUnlocker();
+        $this->experienceAchievementUnlocker = new ExperienceAchievementUnlocker();
+        $this->seriesCompletionAchievementUnlocker = new SeriesCompletionAchievementUnlocker();
+    }
+
+    public function add(AddExperience $command): self
+    {
+        $previousCount = $this->experienceCount;
+
+        $this->recordThat(new ExperienceEarned(
+            userId: $command->userId,
+            amount: $command->amount,
+        ));
+
+        $currentCount = $this->experienceCount;
+
+        $achievement = $this->experienceAchievementUnlocker->achievementToBeUnlocked(
+            previousCount: $previousCount,
+            currentCount: $currentCount,
+            userId: $command->userId
+        );
+
+        if ($achievement) {
+            $this->unlockAchievement(new UnlockAchievement(
+                uuid: $this->uuid(),
+                userId: $command->userId,
+                achievement: $achievement,
+            ));
+        }
+
+        return $this;
+    }
+
+    protected function applyExperienceEarned(ExperienceEarned $event): void
+    {
+        $this->experienceCount += $event->amount;
+    }
+
+    public function unlockAchievement(UnlockAchievement $command): self
+    {
+        $this->recordThat(new AchievementUnlocked(
+            userId: $command->userId,
+            achievementId: $command->achievement->id,
+            slug: $command->achievement->slug,
+            title: $command->achievement->title,
+            description: $command->achievement->description,
+        ));
+
+        return $this;
+    }
+
+    public function registerPullRequest(RegisterPullRequest $command): self
+    {
+        $this->recordThat(new PullRequestMerged(
+            userId: $command->userId,
+            reference: $command->reference,
+        ));
+
+        $achievement = $this->pullRequestAchievementUnlocker->achievementToBeUnlocked(
```

Comment:
I could refactor achievement triggering to separate reactors. It would keep these AR methods a little more clean

Reply:
The downside is: more separate classes. Personally, I'm not against that

Reply:
@freekmurze what's your opinion?